jni/util/vtx_manager.c

```diff
@@ -0,0 +1,73 @@
+#include <stdlib.h>
+#include <dlfcn.h>
+#include "json/osd_config.h"
+#include "util/debug.h"
+#include "vtx_manager.h"
+
+#define VTX_MPS_CONFIG_KEY "vtx_msp"
+#define CHANNEL_PUBLIC 8
+
+static void *tp1801_gui_lib = NULL;
+static uint32_t (* setChannelPilotOriginal)(void *this,unsigned short param_1, bool param_2) = 0;
+static uint32_t (* userSettingsGetInstanceOriginal)() = 0;
+static uint32_t userSettingsInstance = 0;
+static __gs_gui_config *gs_gui_config = 0;
+static int8_t currentChannel = -1;
+
+void setupManager() {
+    //Load SetPilotChannel original
+    setChannelPilotOriginal = dlsym (RTLD_NEXT, "_ZN17GlassUserSettings15setPilotChannelEtb");
+    if (setChannelPilotOriginal == NULL) {
+        if(!tp1801_gui_lib){
+            tp1801_gui_lib = dlopen("/system/lib/libtp1801_gui.so", 1);
+        }
+        setChannelPilotOriginal = dlsym (tp1801_gui_lib, "_ZN17GlassUserSettings15setPilotChannelEtb");
+        if (setChannelPilotOriginal == NULL)
+        {
+            printf("dlsym: %s\n", dlerror());
+            return;
+        }
+    }
+
+    //Load UserSettingGetInstance original
+    userSettingsGetInstanceOriginal = dlsym (tp1801_gui_lib, "_ZN17GlassUserSettings11getInstanceEv");
+    if (userSettingsGetInstanceOriginal == NULL)
+    {
+        printf("dlsym: %s\n", dlerror());
+        return;
+    }
+
+    userSettingsInstance = userSettingsGetInstanceOriginal();
+    if(userSettingsInstance == NULL) {
+        printf("dlsym: error loading user setting instance");
+        return;
+    }
+    gs_gui_config = (__gs_gui_config *)*(uint32_t *)((int)userSettingsInstance + 0xe4);
+}
+
+void changeChannel(int8_t channel) {
+    if(!get_boolean_config_value(VTX_MPS_CONFIG_KEY)) {
```

Comment:
Additional sanity checks are not bad but in my opinion if we consider "separation of concerns" then this file should be as purely as shim to get a clean interface to the GUI functions as possible. Therefor it should not care about msp-osd config state and assume that if it's being called it should do it's job.

The check should be in the overlay itself.

Reply:
Agree, I will move the config check to the caller class place.